RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+
+    //if (m_commonParamsGPU.theThickness!=p.theThickness)
+    //  std::cout << i << (g.isBarrel ? "B " : "E ") << m_commonParamsGPU.theThickness<<"!="<<p.theThickness << std::endl;
+
+    auto ladder = ttopo_.pxbLadder(p.theDet->geographicalId());
+    if (oldLayer != g.layer) {
+      oldLayer = g.layer;
+      // std::cout << "new layer at " << i << (g.isBarrel ? " B  " :  (g.isPosZ ? " E+ " : " E- ")) << g.layer << " starting at " << g.rawId << std::endl;
+      // std::cout << "old layer had " << nl << " ladders" << std::endl;
+      nl = 0;
+    }
+    if (oldLadder != ladder) {
+      oldLadder = ladder;
+      // std::cout << "new ladder at " << i << (g.isBarrel ? " B  " :  (g.isPosZ ? " E+ " : " E- ")) << ladder << " starting at " << g.rawId << std::endl;
+      // std::cout << "old ladder ave z,r,p mz " << zl/8.f << " " << rl/8.f << " " << pl/8.f  << ' ' << miz << ' ' << mxz << std::endl;
+      rl = 0;
+      zl = 0;
+      pl = 0;
+      miz = 90;
+      mxz = 0;
+      nl++;
+    }
+
+    g.shiftX = 0.5f * p.lorentzShiftInCmX;
+    g.shiftY = 0.5f * p.lorentzShiftInCmY;
+    g.chargeWidthX = p.lorentzShiftInCmX * p.widthLAFractionX;
+    g.chargeWidthY = p.lorentzShiftInCmY * p.widthLAFractionY;
+
+    g.x0 = p.theOrigin.x();
+    g.y0 = p.theOrigin.y();
+    g.z0 = p.theOrigin.z();
+
+    auto vv = p.theDet->surface().position();
+    auto rr = pixelCPEforGPU::Rotation(p.theDet->surface().rotation());
+    g.frame = pixelCPEforGPU::Frame(vv.x(), vv.y(), vv.z(), rr);
+
+    zl += vv.z();
+    miz = std::min(miz, std::abs(vv.z()));
+    mxz = std::max(mxz, std::abs(vv.z()));
+    rl += vv.perp();
+    pl += vv.phi();  // (not obvious)
+
+    // errors .....
+    ClusterParamGeneric cp;
+    auto gvx = p.theOrigin.x() + 40.f * m_commonParamsGPU.thePitchX;
+    auto gvy = p.theOrigin.y();
+    auto gvz = 1.f / p.theOrigin.z();
+    //--- Note that the normalization is not required as only the ratio used
+
+    // calculate angles
+    cp.cotalpha = gvx * gvz;
+    cp.cotbeta = gvy * gvz;
+
+    cp.with_track_angle = false;
+
+    auto lape = p.theDet->localAlignmentError();
+    if (lape.invalid())
+      lape = LocalError();  // zero....
+
+#ifdef DUMP_ERRORS
+    auto m = 10000.f;
+    for (float qclus = 15000; qclus < 35000; qclus += 15000) {
+      errorFromTemplates(p, cp, qclus);
+
+      std::cout << i << ' ' << qclus << ' ' << cp.pixmx << ' ' << m * cp.sigmax << ' ' << m * cp.sx1 << ' '
+                << m * cp.sx2 << ' ' << m * cp.sigmay << ' ' << m * cp.sy1 << ' ' << m * cp.sy2 << std::endl;
+    }
+    std::cout << i << ' ' << m * std::sqrt(lape.xx()) << ' ' << m * std::sqrt(lape.yy()) << std::endl;
+#endif
+
+    errorFromTemplates(p, cp, 20000.f);
+    g.pixmx = std::max(0, cp.pixmx);
+    g.sx[0] = cp.sigmax;
+    g.sx[1] = cp.sx1;
+    g.sx[2] = cp.sx2;
+
+    g.sy[0] = cp.sigmay;
+    g.sy[1] = cp.sy1;
+    g.sy[2] = cp.sy2;
+
+    /*
+    // from run1??
+    if (i<96) {
+      g.sx[0] = 0.00120;
```

Comment:
if it is run1, please remove the commented out code

Reply:
Done in https://github.com/cms-patatrack/cmssw/pull/593/commits/9dc393168636f3c4e48986f7739c554f75cdd1c1 .